src/LoRaWANNode.cpp

```diff
@@ -0,0 +1,552 @@
+  * @author  WI6LABS
+  * @version V1.0.0
+  * @date    28-november-2017
+  * @brief   LoRaWAN Arduino API for I-NUCLEO-LRWAN1 shield
+  ******************************************************************************
+  * @attention
+  *
+  * <h2><center>&copy; COPYRIGHT(c) 2017 STMicroelectronics</center></h2>
+  *
+  * Redistribution and use in source and binary forms, with or without modification,
+  * are permitted provided that the following conditions are met:
+  *   1. Redistributions of source code must retain the above copyright notice,
+  *      this list of conditions and the following disclaimer.
+  *   2. Redistributions in binary form must reproduce the above copyright notice,
+  *      this list of conditions and the following disclaimer in the documentation
+  *      and/or other materials provided with the distribution.
+  *   3. Neither the name of STMicroelectronics nor the names of its contributors
+  *      may be used to endorse or promote products derived from this software
+  *      without specific prior written permission.
+  *
+  * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+  * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+  * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+  * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+  * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+  * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+  * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+  * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+  * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+  * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+  *
+  ******************************************************************************
+  */
+
+#include "LoRaWANNode.h"
+#include "lora_driver.h"
+
+
+LoRaWANNodeClass::LoRaWANNodeClass()
+{
+
+}
+
+/*
+ * @brief  Initializes the LoRa module
+ * @param  pointer to the serial instance uses to communicate with the module
+ * @param  LoRaWAN frequency band: 868MHz (EU) or 915MHz (US)
+ * @param  LoRaWAN node class: A, B or C. NOTE: Classes B and C are not available.
+ * @retval Error code: false if initialization failed else true
+ */
+bool LoRaWANNodeClass::begin(HardwareSerial *serialx, uint8_t band, uint8_t loraClass)
+{
+  uint8_t nbTry;
+
+  if(!IS_BAND(band) || !IS_CLASS(loraClass) || (serialx == NULL)) {
+    return 0;
+  }
+
+  if(loraClass != LORA_CLASS_A) {
+    printf("The firmware only supports Class A so far.\r\n");
+    return 0;
+  }
+
+  // Enable UART
+  if(Modem_IO_Init(serialx) != AT_OK) {
+    return 0;
+  }
+
+  LoRa_DumyRequest();
+
+  //Echo mode must be disabled
+  Modem_AT_Cmd(AT_EXCEPT, AT_ATE, 0);
+
+  // Enable Lora module
+  nbTry = 0;
+  while(Lora_Init() != MODULE_READY) {
```

Comment:
Because after disable the echo mode the first AT answer doesn't match with the format expected.

Reply:
This cannot be guessed so please explain this in detail here.

Reply:
Sometimes if the module is not ready when we call `Modem_AT_Cmd(AT_EXCEPT, AT_ATE, 0);` the first answer of the command `Modem_AT_Cmd(AT_CTRL, AT, NULL );` inside `LoraInit()` is not "OK" but "AT OK" (echo of the command sent). The driver doesn't understand the answer in this case. So I used a loop to avoid this issue.

Reply:
When I say "here" I mean in the comments of the file